### ISSUE TITLE
JSON Schema 4.2

### ISSUE THREAD

**Issue by mfenner** [datacite/schema#52]
Provide a JSON schema for the metadata schema 4.2

**Comment by RightInTwo** [comment]
This will be highly appecriated

**Comment by mark-saeon** [comment]
Next step is to align the schema structurally with the example records provided by @mfenner.

This is mostly a matter of updating the schema definition. However, I'd like to make one suggestion for a change to the example records (which I guess implies a change to the JSON structure used internally by DataCite):

The "affiliation" sub-property of the "creator" and "contributor" properties is defined in the DataCite specification as having cardinality of 0-n (supporting the case of an individual having multiple institutional affiliations). However, in the example records it appears as just a string. I would suggest changing this to make it an array of strings. So, for example:

    "creators": [
        {
            ...,
            "affiliations": [
                {
                    "affiliation": "DataCite"
                }
            ]
        }
    ],

Comments?

**Comment by mfenner** [comment]
Good point. Will update the JSON via issue https://github.com/datacite/bolognese/issues/57.

**Comment by mark-saeon** [comment]
@mfenner Could you please provide some clarification on the following:

1. In the example records, there is a "types" field which I see is where the "resourceType" and "resourceTypeGeneral" properties are placed. However, the other properties under "types" (namely schemaOrg, citeproc, bibtex, ris) have no corresponding definitions in the DC 4.2 specification. Do these additional properties need to be in the JSON schema? i.e. Would a person / organisation submitting JSON metadata records be expected to provide values for these properties, and if so is there any documentation on this?

2. Same question about the other additional properties not defined in the DC 4.2 spec - namely id, doi, container, schemaVersion, agency, and state. Are any / all of these properties meant to be provided by someone submitting a metadata record, or are they internally derived by DataCite?

3. The DC spec has "identifier" being a singular property (with sub-property identifierType that may only be "DOI"), but in the example records there is an "identifiers" array, allowing multiple identifiers with different types (e.g. "DOI", "URL", "ISBN"). This makes sense to me, however it's a departure from the 4.2 spec. Should we nevertheless support this "identifiers" array in the JSON schema? If so, should we require at least one of those identifiers to be of type "DOI", or does it not matter? And does this have any bearing on the top-level "id" and "doi" properties?

Thanks!

**Comment by mfenner** [comment]
@mark-saeon 

The `types` other than `resourceTypeGeneral` and `resourceType` are not in the XML schema. They are used for metadata conversion, e.g. converting metadata to RIS that was originally submitted as bibtex. A better solution would be to have the DataCite schema properly support specific work types for text documents, but that has to wait until at least the next major schema update. Interested in your view on how best to support this in JSON Schema.  

The `identifiers` array is a combination of `identifier` and `alternate_identifier` and there is always one of type `DOI`. I found this easier as we already have `id` and `doi`.

As for the other properties:

* id and doi are the same, and are again duplicated in the `identifiers` array. This duplication is for  convenience, as an id is useful in a JSON API and people want easy access to the DOI.
* container is generated from seriesInformation and relatedIdentifier of type isPartOf, so could be left out.
* schemaVersion is from the XML schema.
* agency is the DOI registration agency and currently defaults to "DataCite". In the future this may include "Crossref", as we can easily convert Crossref metadata into this format.
* state is a property of the database and does have to be in the schema. Similar to updated date. The URL and contentURL is not in the XML schema, but I think makes sense to put in the schema.

Happy to discuss and do these things differently.

**Comment by mark-saeon** [comment]
A general point to note about JSON schemas: unlike in XML validation, JSON validators are supposed to ignore anything that they don't recognise. This means a document may include additional elements not defined in the schema, and these won't affect validation.

So, we can use this to our advantage. I'd suggest that (for now) we simply leave the work types out of the schema, until they've been finalised in the next major release. Thus `bibtex` etc may still be included under `types`, but we just won't check those properties, and their presence won't invalidate the entire metadata document.

Regarding identifiers and DOIs (and container and state too, for that matter), I'm not entirely clear on what must be supplied by an end user and what gets generated by DataCite thereafter. For example, if I am submitting metadata, can I leave out `id` and `doi`, and just populate the identifiers array, including one of type "DOI"? Would DataCite then generate the `id` and `doi`? Or is it my responsibility to also set `id` and `doi`? Are `container` and `state` both internally generated, and view-only from the perspective of the user?

In principle I'm trying to ascertain:
- is there a difference between a JSON metadata document as submitted to DataCite, and the same document as retrieved from DataCite?
- if so, can an 'output' document can be resubmitted with modifications as an 'input' document, and can any generated fields (e.g. state) be directly updated by such an input?

I'm asking these questions because I think we should aim to create a schema that is focused on the input, rather than on the internal or output structure of the metadata.

**Comment by mfenner** [comment]
Good points. Several things in the DataCite JSON are read-only, for example `container`, which is generated from `Description` with `DescriptionType` seriesInformation and `relatedIdentifier` of `relationType` isPartOf. `state` is generated internally from an `event` (we are using a state machine, and events are for example `publish` to go from `draft` to `findable`), and `isActive` is set after that. 

Focussing on the input and ignoring extra stuff that is read-only or not part of the schema sounds like a nice way forward.

**Comment by rpwagner** [comment]
@mark-saeon, one point of clarification regarding JSON Schema. If the [`additionalProperties`](`https://json-schema.org/understanding-json-schema/reference/object.html) keyword is set to `false`, then no extra properties are allowed. It can also be used to control the datatype of any additional properties.

**Comment by mark-saeon** [comment]
@rpwagner, thanks for the tip! I'd seen `additionalProperties` but hadn't thought of using it that way.

**Comment by mark-saeon** [comment]
@mfenner, I just want to clarify our approach to specifying the primary identifier / DOI in the schema.

As with the generated fields discussed above, I would suggest that the schema _not_ include the top-level `id` and `doi` fields, since these can be derived from user-supplied info in the `identifiers` array. This also avoids the situation of requiring a user to duplicate information (with possibility of discrepancies).

Would you agree with this? Thus, the schema should only check for an `identifiers` array (which combines primary and alternate ids), and should ensure that the array contains an item with type "DOI".

**Comment by mfenner** [comment]
@mark-saeon I think it depends on how the schema is used. Do we for example think people will use it for other identifiers besides DOIs, as some people are doing with the XML schema? Then what you say makes sense. But the `doi` is really key, and having it in an identifiers array with one element of type `DOI` might create too much overhead. Would it maybe make sense to go back to what the XML schema is doing, i.e. one `identifier` plus multiple `alternateIdentifiers`? And ignoring `doi` and `id`?

**Comment by mfenner** [comment]
BTW, affiliation is now an array, updated yesterday in datacite/bolognese#57.

**Comment by mark-saeon** [comment]
> Do we for example think people will use it for other identifiers besides DOIs, as some people are doing with the XML schema?

Are you saying that some people are using the `identifier` property for non-DOIs?

It's easy enough to implement validation ensuring that an `identifiers` array contains an item with specific content (i.e. a specific `identifierType`), but in principle I think it's better to have separate properties for a primary `identifier` and `alternateIdentifiers`. This way the schema is more resilient to usage changes and we can always be sure how to derive the `id`.

Deriving the `doi` could however be an issue if the `identifier` does not actually contain a DOI, but I'll assume you have some kind of fallback approach in this case (e.g. just copy the identifier as is into the `doi` field). In any case I would agree with ignoring both `id` and `doi` from the perspective of the schema.

**Comment by mark-saeon** [comment]
I noticed a discrepancy between the XML and JSON example records. In the XML the `identifier` is given as a just the DOI string (e.g. "10.5072/example-full"), whereas in the JSON examples it (the item in `identifiers` with type "DOI") is a URL to doi.org (e.g. "https://doi.org/10.5072/example-full").

Is this a mistake in the JSON examples? The specification document also says
> Format should be “10.1234/foo”

**Comment by mark-saeon** [comment]
@mfenner What kind of validation is currently done on `identifier` for uploaded XML metadata? I seem to recall you mentioning in a previous conversation that you are no longer validating that it's a DOI-formatted string, due to usage changes. However, above you've commented that
> the `doi` is really key

Do you mean here "key" as in "important", or as in "primary/unique key" in a database table? :slightly_smiling_face:  I know I'm being pedantic, but I'm trying to understand whether the content of `identifier` can control updates to existing metadata records. e.g. Can one user update metadata that was previously uploaded by a different user, by supplying the same `identifier`? Are there currently any checks in place to ensure that the DOI supplied by an organization has the correct prefix for that organization? If we allow other types of identifier (e.g. ISBN, etc), do we want to check that the supplied id is valid within the given id-space and is valid for that organization?

**Comment by mark-saeon** [comment]
Apologies, I've been missing something in my understanding of the metadata upload process. Been checking out the DataCite MDS API, and I see that the DOI forms part of the path in requests to the API, and authorization (e.g. controlling access to a particular DOI prefix) is handled by the API. So I guess that record selection is determined by the DOI in the request path rather than from any content in the metadata itself.

OK then, I think for now the pragmatic approach will be to keep the JSON schema consistent with the XSD: so, as mentioned above, separate `identifier` and `alternateIdentifiers` properties, and ignore `id` and `doi`.

**Comment by lnielsen** [comment]
I've reviewed the JSONSchema and, in general, it looks very good. I've made a PR #59 (based on Tom's #55) and open issues for the issues I've found:

- #60 Missing lang property on publisher 
- #61 Problem with uniqueItems 
- #62 Issue with additional properties being allowed

One discussion that's perhaps worth taking for v5 of the schema is if we should keep the JSON and XML schema's aligned. I don't have strong opinions about it. I see definitely strong points for keeping the two schema's aligned. On the other hand, the JSON looks very XMLish, and is honestly not easy to use in applications. Example:

```json
{
  "identifier" : {"identifier": "10.1234/foo.bar", "identifierType": "DOI"}
}
```

vs

```json
{
  "id" : "10.1234/foo.bar"
}
```

The impact of having long names and nesteded properties is that the code accessing it, becomes very verbose:

```python 
record["identifier"]["identifier"]
# vs 
record["id"]
```

**Comment by mfenner** [comment]
@lnielsen thanks for the input. There are a few places where the JSON is significantly different from the XML, notably `identifiers`, `types` and `container`. The reasons for that is that we also do metadata conversions to and from other formats, and we would for example loose information if we can't specify the `bibtex` type independently from `resourceTypeGeneral` and `resourceTyoe`.

**Comment by mfenner** [comment]
There is also verbosity in the XML that we might want to avoid (e.g. wrapper elements such as `creators` for `creator`).

**Comment by lnielsen** [comment]
Yep, definitely, all the wrapper elements are annoying to work with.

**Comment by tmorrell** [comment]
Two other things I noticed in the JSON version when converting some of our earlier metadata with the new schema.

use of "affiliation" vs "affiliations" for the array label.  Most of the other arrays are plural.
All the gelolocation point values are strings instead of floats.

**Comment by mark-saeon** [comment]
@tmorrell Well spotted re the geolocation points. Updated the 4.2 and 4.3 schemas and examples (PR68) - now validating lats and lons in the same way as in the XSD.

**Comment by lnielsen** [comment]
See my comment on geolocation points - there are good reasons for them to be strings rather than floats.
